internal/graph/memory_test.go

```diff
@@ -0,0 +1,412 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package graph
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"io"
+	"reflect"
+	"testing"
+
+	"github.com/opencontainers/go-digest"
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+	"oras.land/oras-go/v2/internal/cas"
+	"oras.land/oras-go/v2/internal/descriptor"
+)
+
+/*
+A(manifest)-------------------+
+
+	|                         |
+	|                         |
+	|                         |
+	v                         |
+
+B(manifest)-----------------+ |
+
+	|                       | |
+	|                       | |
+	v                       v v
+
+C(layer)                  D(layer)
+*/
+func TestDeletableMemory_IndexAndRemove(t *testing.T) {
+	testFetcher := cas.NewMemory()
+	testDeletableMemory := NewMemory()
+	ctx := context.Background()
+
+	// generate test content
+	var blobs [][]byte
+	var descs []ocispec.Descriptor
+	appendBlob := func(mediaType string, blob []byte) ocispec.Descriptor {
+		blobs = append(blobs, blob)
+		descs = append(descs, ocispec.Descriptor{
+			MediaType: mediaType,
+			Digest:    digest.FromBytes(blob),
+			Size:      int64(len(blob)),
+		})
+		return descs[len(descs)-1]
+	}
+	generateManifest := func(layers ...ocispec.Descriptor) ocispec.Descriptor {
+		manifest := ocispec.Manifest{
+			Layers: layers,
+		}
+		manifestJSON, err := json.Marshal(manifest)
+		if err != nil {
+			t.Fatal(err)
+		}
+		return appendBlob(ocispec.MediaTypeImageManifest, manifestJSON)
+	}
+	descC := appendBlob("layer node C", []byte("Node C is a layer")) // blobs[0], layer "C"
+	descD := appendBlob("layer node D", []byte("Node D is a layer")) // blobs[1], layer "D"
+	descB := generateManifest(descs[0:2]...)                         // blobs[2], manifest "B"
+	descA := generateManifest(descs[1:3]...)                         // blobs[3], manifest "A"
+	testFetcher.Push(ctx, descA, bytes.NewReader(blobs[3]))
+	testFetcher.Push(ctx, descB, bytes.NewReader(blobs[2]))
+	testFetcher.Push(ctx, descC, bytes.NewReader(blobs[0]))
+	testFetcher.Push(ctx, descD, bytes.NewReader(blobs[1]))
+
+	// make sure that testFetcher works
+	rc, err := testFetcher.Fetch(ctx, descA)
+	if err != nil {
+		t.Errorf("testFetcher.Fetch() error = %v", err)
+	}
+	got, err := io.ReadAll(rc)
+	if err != nil {
+		t.Errorf("testFetcher.Fetch().Read() error = %v", err)
+	}
+	err = rc.Close()
+	if err != nil {
+		t.Errorf("testFetcher.Fetch().Close() error = %v", err)
+	}
+	if !bytes.Equal(got, blobs[3]) {
+		t.Errorf("testFetcher.Fetch() = %v, want %v", got, blobs[4])
+	}
+
+	// index test content into graph.DeletableMemory
+	testDeletableMemory.Index(ctx, testFetcher, descA)
+	testDeletableMemory.Index(ctx, testFetcher, descB)
+	testDeletableMemory.Index(ctx, testFetcher, descC)
+	testDeletableMemory.Index(ctx, testFetcher, descD)
```

Comment:
Should we verify the status right after each index operation?

Reply:
good idea.

Reply:
Took me so long to rewrite all the test cases.....